Added terasort test for soak cluster

susanxhuynh opened the pull request:
This test runs teragen, terasort, and teravalidate. Because the soak cluster is long-running, we have to delete the output files from HDFS after each test run. This is done by the hdfs-delete-terasort-files job.
I added a "DO_TEARDOWN" env to indicate that packages are not to be uninstalled in soak cluster testing.

mgummelt commented on tests/test_spark.py:
```diff
@@ -169,6 +216,17 @@ def _require_spark():
     _wait_for_spark()
 
 
+def _require_spark_cli():
```
Is this needed for the soak tests?  I'd rather have the test runner code be responsible for setting up the environment (incuding the CLI), rather than the tests themselves.

mgummelt replied:
Oh, I see.  This function is necessary because `_require_spark` will not install the CLI if the marathon app for the package is already installed then, right?  If so, can you just move this functionality to _require_spark()?  I'd rather have that function just be fully responsible for installing the entire package, including the CLI.

mgummelt commented on tests/test_spark.py:
```diff
@@ -35,19 +36,23 @@ def _init_logging():
 HDFS_SERVICE_NAME='hdfs'
 SPARK_PACKAGE_NAME='spark'
+TERASORT_JAR='https://downloads.mesosphere.io/spark/examples/spark-terasort-1.0-jar-with-dependencies_2.11.jar'
+TERASORT_MAX_EXECUTOR_CORES=6
```
Can we call this `TERASORT_MAX_CORES` instead?  It reads as if it's the max number of cores a single executor will have, but that's not the effect.

mgummelt commented on tests/test_spark.py:
```diff
@@ -66,11 +71,53 @@ def test_jar():
+@pytest.mark.soak
+def test_terasort():
+    if _hdfs_enabled():
+        _delete_hdfs_terasort_files()
```
nice organization!

mgummelt commented on tests/dcosjob.py:
```diff
@@ -0,0 +1,37 @@
+    jobs_folder = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), 'jobs', 'json'
+    )
+    _run_cli('job add {}'.format(
```
Is there no `dcos` or `shakedown` function for this?

mgummelt commented on tests/dcosjob.py:
```diff
@@ -0,0 +1,37 @@
+import os
```
Ideally, this file wouldn't exist, and we would be calling into `shakedown` or `dcos` for all these functions.

mgummelt replied:
What's the underlying function that the dcos cli is using to check that a job is finished?  If that exists in the `dcos` package, let's use that.  Otherwise, CLI is fine.

mgummelt commented on tests/test_spark.py:
```diff
@@ -66,11 +71,53 @@ def test_jar():
+        _run_teragen()
+
+
+def _run_teragen():
```
I think all the other private functions are at the bottom of this file.  Can we maintain that organization?

Actually, I think ideally, the terasort test would be in a separate `test_soak.py` file.  Then the tests would be more organized by their application.

susanxhuynh replied:
I'll put the terasort test in a separate file.

susanxhuynh commented:
@mgummelt Addressed your review comments:
- I reorganized the Terasort code into a separate file, test_soak.py.
- Added a new "utils.py" file for common utility functions.

susanxhuynh commented on tests/test_soak.py:
```diff
@@ -0,0 +1,109 @@
+    shakedown.wait_for(
+        lambda: (
+            'Successful runs: 1' in
+            _run_cli('job history {}'.format(job_name))
```
Checking a job's completion status is messy in the CLI; it hits a `jobs` endpoint and parses the results: https://github.com/dcos/dcos-cli/blob/0.4.16/cli/dcoscli/job/main.py#L266 I will use the CLI for now.

mgummelt replied:
:+1:

susanxhuynh commented on tests/test_soak.py:
```diff
@@ -0,0 +1,109 @@
+        _run_teravalidate()
+
+
+def _require_spark_cli():
```
Do you still want to combine this method with require_spark(), which is in another file?

mgummelt replied:
yes please
